File: go/misc/huawei/q1-search-tree/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type TreeNode struct {
	Val         int
	Left, Right *TreeNode
}

func main() {
	// file, err := os.Open("./testcases.txt")
	// if err != nil {
	// 	fmt.Println("Error", err)
	// 	return
	// }
	//
	// in := bufio.NewReader(file)
	// defer file.Close()
	in := bufio.NewReader(os.Stdin)

	var nums []int
	input, _ := in.ReadString('\n')
	strs := strings.Fields(input)
	for _, str := range strs {
		cur, _ := strconv.Atoi(str)
		nums = append(nums, cur)
	}
	var k int
	fmt.Fscan(in, &k)

	// 1. sort
	sort.Ints(nums)
	// fmt.Println(nums)

	// 2. build
	root := buildBST(nums)

	// 3. check
	fmt.Printf("%v", "S")
	check(root, k)
}

func buildBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	mid := len(nums) / 2
	root := &TreeNode{Val: nums[mid]}

	root.Left = buildBST(nums[:mid])
	root.Right = buildBST(nums[mid+1:])
	return root
}

func check(root *TreeNode, target int) {
	if root == nil {
		fmt.Printf("%v", "N")
		return
	}

	cur := root.Val
	// 如果目标值大于根节点
	if target > cur {
		if root.Right != nil {
			fmt.Printf("%s", "R")
		}
		check(root.Right, target)
	} else if target == cur {
		fmt.Printf("%s", "Y")
	} else if root.Left != nil {
		if root.Left != nil {
			fmt.Printf("%s", "L")
		}
		check(root.Left, target)
	}
}

File: go/misc/huawei/q1-search-tree/cpp/main.cpp
#include <algorithm>
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

TreeNode* buildBST(vector<int>& nums, int start, int end) {
    if (start > end) return NULL;

    int mid = start + (end - start) / 2;
    TreeNode* root = new TreeNode(nums[mid]);

    root->left = buildBST(nums, start, mid - 1);
    root->right = buildBST(nums, mid + 1, end);

    return root;
}

void check(TreeNode* root, int target) {
    if (!root) {
        cout << "N";
        return;
    }

    if (target > root->val) {
        if (root->right) cout << "R";
        check(root->right, target);
    } else if (target == root->val) {
        cout << "Y";
    } else {
        if (root->left) cout << "L";
        check(root->left, target);
    }
}

int main() {
    vector<int> nums;
    string line;
    getline(cin, line);
    istringstream is(line);
    int num;
    while (is >> num) {
        nums.push_back(num);
    }

    int k;
    cin >> k;

    sort(nums.begin(), nums.end());

    TreeNode* root = buildBST(nums, 0, nums.size() - 1);

    cout << "S";
    check(root, k);

    return 0;
}
